README.md

```diff
@@ -5,7 +5,211 @@
+- `coeff_func`: An operational argument for a coefficient function `f(du,u,p,t)`
+  which sets the coefficients of the operator. If `coeff_func` is a `Number`
+  then the coefficients are set to be constant with that number. If `coeff_func`
+  is an `AbstractArray` with length matching `len`, then the coefficients are
+  constant but spatially dependent.
+
+`N`-dimensional derivative operators need to act against a value of at least
+`N` dimensions.
+
+### Example
+
+The 3-dimensional Laplacian is created by:
+
+```julia
+N = 64
+Dxx = CenteredDifference(2,2,dx,N)
+Dyy = CenteredDifference{2}(2,2,dx,N)
+Dzz = CenteredDifference{3}(2,2,dx,N)
+L = Dxx + Dyy + Dzz
+
+u = rand(N,N,N)
+L*u
+```
+
+### Derivative Operator Actions
+
+These operators are lazy, meaning the memory is not allocated. Similarly, the
+operator actions `*` can be performed without ever building the operator
+matrices. Additionally, `mul!(y,L,x)` can be performed for non-allocating
+applications of the operator.
+
+### Concretizations
+
+The following concretizations are provided:
+
+- `Array`
+- `SparseMatrixCSC`
+- `BandedMatrix`
+- `BlockBandedMatrix`
+
+Additionally, the function `sparse` is overloaded to give the most efficient
+matrix type for a given operator. For one-dimensional derivatives this is a
+`BandedMatrix`, while for higher dimensional operators this is a `BlockBandedMatrix`.
+The concretizations are made to act on `vec(u)`.
+
+## Atomic Boundary Value Operators
+
+Boundary conditions are implemented through a ghost node approach. The discretized
+values `u` should be the interior of the domain so that, for the boundary value
+operator `Q`, `Q*u` is the discretization on the closure of the domain. By
+using it like this, `L*Q*u` is the `NxN` operator which satisfies the boundary
+conditions.
+
+### Periodic Boundary Conditions
+
+The constructor `PeriodicBC` provides the periodic boundary condition operator.
+
+### Robin Boundary Conditions
+
+The variables in l are `[αl, βl, γl]`, and correspond to a BC of the form
+`al*u(0) + bl*u'(0) = cl`, and similarly `r` for the right boundary
+`ar*u(N) + br*u'(N) = cl`.
+
+```julia
+RobinBC(l::AbstractArray{T}, r::AbstractArray{T}, dx::AbstractArray{T}, order = one(T))
+```
+
+Additionally, the following helpers exist for the Neumann `u'(0) = α` and
+Dirichlet `u(0) = α` cases.
+
+```julia
+Dirichlet0BC(T::Type)
+DirichletBC(α::AbstractVector{T}, dx::AbstractVector{T}, order = 1)
+Neumann0BC(dx::Union{AbstractVector{T}, T}, order = 1)
+NeumannBC(α::AbstractVector{T}, dx::AbstractVector{T}, order = 1)
+```
+
+### General Boundary Conditions
+
+Implements a generalization of the Robin boundary condition, where α is a vector
+of coefficients. Represents a condition of the form
+α[1] + α[2]u[0] + α[3]u'[0] + α[4]u''[0]+... = 0
+
+```julia
+GeneralBC(αl::AbstractArray{T}, αr::AbstractArray{T}, dx::AbstractArray{T}, order = 1)
+```
+
+### Multidimensional Boundary Conditions
+
+```julia
+Q_dim = MultiDimBC(Q, size(u), dim)
```

Comment:
To be consistent with the derivative operators, this should be `MultiDimBC{dim}(Q, size(u))` @xtalax